### ISSUE TITLE
Question: npx bddgen is not getting exited properly. 

### ISSUE THREAD

**Issue by dpjp** [vitalets/playwright-bdd#197]
Hi @vitalets - Good Day. Am trying to execute the npx bddgen with the below version. I see that the features-gen folder is successfully created with test *.spec.js files, but the process is not getting exited properly. Not sure what causes this problem. In  local am able to do "Ctrl + C" and continue with my execution. But in pipeline, this particular step `npx bddgen` is taking longer time and its not getting exited. Also, tried debugging with nodemon mentioned in the documentation, still no luck. Can you help me here how can I fix this issue ? 
Attached is the screenshot for your reference. 

Please let me know if you need any other information. 
![pw-npx-bddgen-ExitIssue](https://github.com/user-attachments/assets/8a12ff44-7063-4316-a41b-0cc3d347c91c)

**Note**: tried clearing npm cache, uninstall and reinstall playwright-bdd,

**Comment by vitalets** [comment]
Hi @dpjp !
1. Could you run locally `npx bddgen --verbose` and share the output?
2. Could you clone [playwright-bdd-example](https://github.com/vitalets/playwright-bdd-example) and run `npx bddgen` there?

**Comment by dpjp** [comment]
Hi @vitalets ,

I tried the above commands. The npx bddgen works fine in the playwright-bdd-example project. And alsom observed that the issue was causing because of aspose.cells nodejs via java npm package. Am using this package in my project for excel manipulations. 

**Aspose.cells npm package**
https://www.npmjs.com/package/aspose.cells

If I have aspose.cells installed in my project, the npx bddgen is not exiting properly. But after uninstalling the aspose.cells, I see that the npx bddgen seems to be working fine. 
Aspose,cells is must library for my project as we deal with different excel templates. It would be great if you could help me with a work around.

**Comment by vitalets** [comment]
I've tried to install `aspose.cells` but getting an error (osx, m1):
```
> npm i aspose.cells

...
../src/java.h:7:10: fatal error: 'jni.h' file not found
```

But I see in readme they highly recommend to migrate to [Aspose.Cells for Node.js via C++](https://www.npmjs.com/package/aspose.cells.node). I've checked - it installs correctly and `npx bddgen` works.
Could try with it? 
Looks like support for `aspose.cells` will be dropped anyway in future.

**Comment by dpjp** [comment]
Hi @vitalets @VitaliyPotapov - For the past few days, I had been following up on this with aspose.cells support and my project engineering team. Due to security concerns and performance issues, I will not be able to use aspose.cells for Node.js via c++ as they set license using a file and cannot have the license file in shared repository, whereas, for nodejs via java we use metered license and we can store the keys in secrets.  Also, aspose support team confirmed that they will be supporting this aspose.cells for nodejs via Java in the future and dont have any plans to retire/deprecate it.  So I think we have only this option(use aspose.cells for nodejs via Java) to go with. 

Aspose support ticket reference: https://forum.aspose.com/t/aspose-cells-node-version-24-7-npm-library/288564

Issue Description:
I can say exactly when this npx bddgen is not exiting properly.

I was able to do npm install for aspose.cells and Java. 
I installed Java JDK(v22) in my machine. 

**After the above installation steps, npx bddgen works fine.**

In order to use the aspose.cells for nodejs via java in my office laptop, I am forced to set the certificate in Java bin folder using keytool command as mentioned below 

**"keytool -import -trustcacerts -file <Path to Certificate>/ca-certificatename -alias custom-Root-CA -keystore ..\jre\lib\security\cacerts"**

Only after setting the certificate, am facing this issue.(npx bddgen is not exiting properly).

Can you help me if you have any workaround towards this ?

**Comment by vitalets** [comment]
Hey @dpjp 
Such issue is very difficult to reproduce.

Lets try the following:
1. When you run pure Playwright tests, do es it exit correctly with and without installed certs?
2. Try to run this command with and without installed certs, to check what causes node to hang (uses this [module](https://www.npmjs.com/package/why-is-node-running)):
   ```
   npx why-is-node-running node_modules/playwright-bdd/dist/cli/index.js --verbose
   ```

**Comment by dpjp** [comment]
Hi @vitalets, @VitaliyPotapov - Sorry.. Ignore my previous recent post. I see the issue is in the below lines. In one of my .ts class, 
 am importing the aspose.cells using the below lines. 
 
var aspose:any = aspose || {};
aspose.cells= require("aspose.cells");

![npx_bddgen_rootcause_AsposeCells](https://github.com/user-attachments/assets/62f7c269-e86e-4c22-af2a-7e7a3ad52ace)

The same code works with playwright test runner(excluded playwright-bdd).  Do you have any work around for this ?

**Comment by vitalets** [comment]
If you just run the following file with `node`, does it exit properly?
```js
// file.js
var aspose = aspose || {};
aspose.cells= require("aspose.cells");
```
Run:
```
node file.js
```

**Comment by vitalets** [comment]
Ok. Then please try this command, to get the exact reason of hanging:
```
npx why-is-node-running file.js
```

Also, haven't you open an issue in [aspose forum](https://forum.aspose.com)? Because hanging of `require("aspose.cells");` looks like a more general issue than related to playwright-bdd.

In Playwright it exists, because Playwright internally calls `process.exit()` to force exiting.

**Comment by dpjp** [comment]
Hi @vitalets - I tried running this command 'npx why-is-node-running file.js' and I see the below error. 

![image](https://github.com/user-attachments/assets/51b3a614-ebc6-4b09-8d3c-9c8e63e3c3ff)

Yes, I did posted in the aspose.cells forum and I got the below response and still no luck. 

![image](https://github.com/user-attachments/assets/aee9b86f-d9c5-47a7-b649-9d91394ac278)

**Comment by vitalets** [comment]
Then we can try to wrap the call of `bddgen` and force exit when files are generated:
// bddgen.cjs
```js
const { spawn } = require('node:child_process');
const child = spawn('./node_modules/.bin/bddgen', ['--verbose']); // for windows there can be bddgen.bat

child.stdout.on('data', (data) => {
    console.log(`stdout: ${data}`);
    if (data.includes('Generating Playwright tests')) {
        setTimeout(() => child.kill(), 2000);
    }
});
```
And run it: 
```
node bddgen.cjs
```
Let me know, does it work?

**Note:**
Now it uses timeout for 2000 ms.
This could be improved in the next version of pw-bdd, I will add a string `Done.` as a final output to let child process know that files are generated.

**Comment by dpjp** [comment]
Hi @vitalets  - Am I doing anything wrong here? Tried executing the same. I get the below error and says that its unable to locate the executable. 

![image](https://github.com/user-attachments/assets/f91547a2-0504-4797-8585-151109b09ff2)

Also, in the bin folder under node modules, I see the below files. 

![image](https://github.com/user-attachments/assets/8f8d1dd3-42ed-439d-99c5-24bb3c0dbd97)

**Comment by vitalets** [comment]
Hi @dpjp 
Could you try with `const child = spawn('./node_modules/.bin/bddgen.cmd', ['--verbose']);`

**Comment by dpjp** [comment]
Hi @vitalets  - Yes. I did try the same previously. I was getting the below error. 

![image](https://github.com/user-attachments/assets/7fc064a7-40a7-45ec-8381-9c848d8f645a)

Also, wanted to check on the below. I had generated the steps definitions for these steps. But still it says undefined steps and when pressed "F12" from the feature file, Its not navigating to the corresponding step definition. I tried setting the cucumber. Glue in settings.json file as well. 

![image](https://github.com/user-attachments/assets/4183883c-bcce-491c-bd33-701080cf7b24)

**Comment by vitalets** [comment]
> I did try the same previously for .cmd and .ps1. I was getting the below error.

I'm not on windows, so tried to ask chatgpt to check, that's the suggested code:
```js
const { spawn } = require('child_process');
const path = require('path');

// Assuming `bddgen` is installed as a dependency in your project
const bddgenPath = path.resolve(__dirname, 'node_modules', '.bin', 'bddgen');

const child = spawn(bddgenPath, ['--verbose'], {
  stdio: 'inherit',  // This option will output the child process logs in the parent process
  shell: true        // Helps with path resolution on Windows
});

child.on('error', (err) => {
  console.error('Failed to start subprocess:', err);
});

child.on('close', (code) => {
  console.log(`Child process exited with code ${code}`);
});
```
Please try with it.

> Also, wanted to check on the below...

Lets track it in a separate issue. Please share project structure or try to reproduce it based on https://github.com/vitalets/playwright-bdd-example

**Comment by vitalets** [comment]
Version 8.0.1 shipped improved output for `npx bddgen`. Now it prints `Done` when all files have been generated.
It allows to detect end of process with third-party tools.

Example output:
```
$ npx bddgen --verbose

Loading features: features/*.feature
Found feature files: 1
  - features/homepage.feature
Loading steps: features/steps/*.ts
Found step files: 2
  - features/steps/fixtures.ts (0 steps)
  - features/steps/index.ts (3 steps)
Clearing output dir: .features-gen/**/*.spec.js
Generating Playwright test files: 1
  - .features-gen/features/homepage.feature.spec.js
Done (0.5s).
```
